components/hybrid_engine/src/observer/load_eviction.rs

```diff
@@ -242,18 +248,83 @@ impl AdminObserver for LoadEvictionObserver {
     fn pre_transfer_leader(
         &self,
         ctx: &mut ObserverContext<'_>,
         _tr: &kvproto::raft_cmdpb::TransferLeaderRequest,
-    ) -> raftstore::coprocessor::Result<Option<kvproto::raft_serverpb::ExtraMessage>> {
-        if !self.cache_engine.region_cached(ctx.region()) {
+    ) -> raftstore::coprocessor::Result<Option<TransferLeaderCustomContext>> {
+        // Warm up transferee's cache when a region is in active or in loading.
+        let active_only = false;
+        if !self.cache_engine.region_cached(ctx.region(), active_only) {
             return Ok(None);
         }
-        let mut msg = ExtraMessage::new();
-        msg.set_type(ExtraMessageType::MsgPreLoadRegionRequest);
-        Ok(Some(msg))
+        let mut value = vec![];
+        value
+            .write_var_i64(ExtraMessageType::MsgPreLoadRegionRequest.value() as i64)
+            .unwrap();
+        Ok(Some(TransferLeaderCustomContext {
+            key: TRANSFER_LEADER_CONTEXT_KEY.to_vec(),
+            value,
+        }))
+    }
+
+    fn pre_ack_transfer_leader(
+        &self,
+        r: &mut ObserverContext<'_>,
+        msg: &raft::eraftpb::Message,
+    ) -> bool {
+        fn get_value(ctx: &[u8]) -> raftstore::Result<Option<ExtraMessageType>> {
+            let ctx = TransferLeaderContext::from_bytes(ctx)?;
+            let Some(mut value) = ctx.get_custom_ctx(TRANSFER_LEADER_CONTEXT_KEY) else {
+                return Ok(None);
+            };
+            let value = decode_var_i64(&mut value)?;
+            Ok(ExtraMessageType::from_i32(value as i32))
+        }
+
+        let region = r.region();
+        let context = msg.get_context();
+        let ty = match get_value(context) {
+            Ok(Some(ty)) => ty,
+            other => {
+                // For compatibility, return ready if the context is not found
+                // or invalid.
+                info!("ime transfer leader context not found";
+                    "region_id" => ?region.get_id(),
+                    "from" => ?msg.get_from(),
+                    "error" => ?other.err());
+                return true;
+            }
+        };
+
+        let need_warmup = ty == ExtraMessageType::MsgPreLoadRegionRequest;
+        if !need_warmup {
+            // Ready to ack.
+            return true;
+        }
+
+        if region.get_peers().is_empty() {
+            // MsgPreLoadRegionRequest is sent before leader issue a transfer leader
+            // request. It is possible that the peer is not initialized yet.
+            warn!("ime skip warmup an uninitialized region"; "region" => ?region);
+            return true;
+        }
+
+        // Exclude loading states to make sure the region is active.
+        let active_only = true;
+        let has_cached = self.cache_engine.region_cached(r.region(), active_only);
+        if has_cached {
+            // Ready to ack.
+            return true;
+        }
+
+        self.cache_engine.load_region(r.region());
```

Comment:
This triggers loading only when the ticker ticks, can we make it more actively?

Reply:
The loading is initiated[^1] when a peer receives a MsgTransferLeader and a transferee checks the result on every tick.

[^1]: https://github.com/overvenus/tikv/blob/2009bf4dd16bbdd50a91f7ecf76bf827d79a437c/components/raftstore/src/store/peer.rs#L4834